File: app/src/main/java/com/faendir/lightning_launcher/multitool/proxy/JavaScript.java
package com.faendir.lightning_launcher.multitool.proxy;

import android.content.Intent;
import android.view.View;
import androidx.annotation.Keep;
import androidx.annotation.NonNull;

/**
 * @author lukas
 * @since 09.07.18
 */
public interface JavaScript {
    @Keep
    interface CreateMenu extends JavaScript {
        default void showMenu(Object jsMenu, Object jsItem) {
            showMenu(ProxyFactory.lightningProxy(jsMenu, com.faendir.lightning_launcher.multitool.proxy.Menu.class), ProxyFactory.lightningProxy(jsItem, Item.class));
        }

        void showMenu(com.faendir.lightning_launcher.multitool.proxy.Menu menu, Item item);
    }

    @Keep
    interface ActivityResult extends JavaScript {
        void onActivityResult(int resultCode, Intent data, String token);
    }

    @Keep
    interface CreateCustomView extends JavaScript {
        default View onCreate(Object jsItem) {
            return onCreate(ProxyFactory.lightningProxy(jsItem, CustomView.class));
        }

        View onCreate(CustomView item);
    }

    @Keep
    interface Setup extends JavaScript {
        void setup();
    }

    @Keep
    interface Normal extends JavaScript {
        void run();
    }

    @Keep
    interface Direct extends JavaScript {
        String PARAM_CLASS = "multitool$classname";
        String PARAM_DATA = "multitool$data";

        @NonNull
        String execute(String data);
    }

    @Keep
    interface Listener extends JavaScript {
        void register();

        void unregister();

        default void handleCommand(String command) {

        }
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/proxy/JavaScript.kt
package com.faendir.lightning_launcher.multitool.proxy

import android.content.Intent
import android.view.View
import androidx.annotation.Keep

/**
 * @author lukas
 * @since 09.07.18
 */
interface JavaScript {
    @Keep
    interface CreateMenu : JavaScript {
        fun showMenu(jsMenu: Any, jsItem: Any) {
            showMenu(ProxyFactory.lightningProxy(jsMenu, Menu::class.java), ProxyFactory.lightningProxy(jsItem, Item::class.java))
        }

        fun showMenu(menu: Menu, item: Item)
    }

    @Keep
    interface ActivityResult : JavaScript {
        fun onActivityResult(resultCode: Int, data: Intent, token: String)
    }

    @Keep
    interface CreateCustomView : JavaScript {
        fun onCreate(jsItem: Any): View {
            return onCreate(ProxyFactory.lightningProxy(jsItem, CustomView::class.java))
        }

        fun onCreate(item: CustomView): View
    }

    @Keep
    interface Setup : JavaScript {
        fun setup()
    }

    @Keep
    interface Normal : JavaScript {
        fun run()
    }

    @Keep
    interface Direct : JavaScript {

        fun execute(data: String): String

        companion object {
            const val PARAM_CLASS = "multitool\$classname"
            const val PARAM_DATA = "multitool\$data"
        }
    }

    @Keep
    interface Listener : JavaScript {
        fun register()

        fun unregister()

        fun handleCommand(command: String) {}
    }
}
